Web/Controller/Tasks.hs

```diff
@@ -0,0 +1,203 @@
+module Web.Controller.Tasks where
+
+import Web.Controller.Prelude
+import Web.View.Tasks.Index
+import Web.View.Tasks.Form
+
+-- Decode data coming from frontend.
+import Data.Aeson.Types (Parser)
+import qualified Data.Aeson (Value, toJSON, (.=))
+import Network.Wai (Request, getRequestBodyChunk)
+
+-- Allow encoding HTML so we can handover to the frontend as JSON.
+import qualified Text.Blaze.Html.Renderer.Text as Renderer
+import Text.Blaze.Html (Html)
+import qualified Data.Text.Lazy as TL
+
+-- For creating the Tree structure.
+import Data.Tree (Tree(..))
+import qualified Data.Tree as Tree
+import Data.Tree.Zipper
+
+
+
+instance Controller TasksController where
+
+    action UpdateSortTasksAction = do
+            decodedTree <- decodeRequestTree request
+            case decodedTree of
+                Right uuidTree -> do
+                    tasks <- query @Task |> fetch
+
+                    -- Go over the tree and update the weights and the reference to the parent,
+                    -- if there is one.
+                    -- If the level is 0, then we should not reference any parent.
+                    -- They way we should do it, is by going over the tree with a zipper
+                    -- so we can know the level and weight of each element, and the
+                    updateFromZipper tasks zipper Nothing 0
+                    redirectTo TasksAction
```

Comment:
@mpscholten After sorting a task, and making it a child of another task --  the parent task should not longer have the "Delete" link.

But seems that the redirectTo doesn't trigger a refresh of the page. If you will try to reload the page, you'll see it correctly. Is that an issue with turbolinks?


https://github.com/amitaibu/ihp-sortable-tree/assets/125707/3274a585-e2c0-4218-87e7-86ab72bbe44a

Reply:
This is not realted to turbolinks. Right now it's not involved at all, as you're doing your own ajax request.

In the `app.js` you're manually removing the delete link:

```js
 $(targetParentNode).find('.js-delete:first-child').remove();
```

The problem here is that `$(targetParentNode).find('.js-delete:first-child')` doesn't match any dom element. I've found this fix to work well:

```js
$(targetParentNode).find('.js-delete').first().remove();
```

Alternatively you can also enable turbolinks here by calling IHP's global `transitionToNewPage` with the HTML of the response.

```js
                    const response = await fetch('/UpdateSortTasks', {
                        method: 'POST',
                        headers: {
                            'Content-Type': 'application/json'
                        },
                        body: JSON.stringify({ tree: filteredNodes })
                    })

                    if (!response.ok) {
                        throw new Error(`HTTP error! Status: ${response.status}`);
                    }

                    const html = await response.text(); // Get the HTML from the response
                    var parser = new DOMParser();
                    var doc = parser.parseFromString(html, 'text/html'); // turn the HTML into a dom node

                    transitionToNewPage(doc); // This will replace the page's html using morphdom
```

After that the call to `const tree = new SortableTree({` needs to be done again because the old dom nodes have been replaced. For that I wrapped the init callback into a variable `initHandler` and then call the `initHandler` after `transitionToNewPage`:

```js
const initHandler = function () {
    const container = document.getElementById('treeContainer');
    if (!!container) {
        const nodes = JSON.parse(container.getAttribute('data-tree'));

        const tree = new SortableTree({
            nodes: nodes,
            initCollapseLevel: 10,
            element: document.querySelector('#treeContainer'),
            renderLabel: (data) => {
                return `<span>${data.body}</span>`;
            },

            onChange: async ({ nodes, movedNode, srcParentNode, targetParentNode }) => {
                const filterNodes = function(nodes) {
                    return nodes.map(node => {
                        let newNode = {
                            element: {
                                _data: {
                                    uuid: node.element._data.uuid
                                }
                            }
                        };

                        if (node.subnodes && node.subnodes.length > 0) {
                            newNode.subnodes = filterNodes(node.subnodes);
                        }
                        else  {
                            newNode.subnodes = [];
                        }

                        return newNode;
                    });
                }
                const filteredNodes = filterNodes(nodes);

                try {
                    const response = await fetch('/UpdateSortTasks', {
                        method: 'POST',
                        headers: {
                            'Content-Type': 'application/json'
                        },
                        body: JSON.stringify({ tree: filteredNodes })
                    })

                    if (!response.ok) {
                        throw new Error(`HTTP error! Status: ${response.status}`);
                    }

                    const html = await response.text();
                    var parser = new DOMParser();
                    var doc = parser.parseFromString(html, 'text/html');

                    transitionToNewPage(doc);
                    initHandler();

                    // Remove the Delete from the parent, but not from the child.
                    //$(targetParentNode).find('.js-delete').first().remove();

                } catch (error) {
                    console.error("There was a problem with the fetch operation:", error.message);
                }
                },
                onClick: (event, node) => {
                    // console.log(node.data);
                },
            });

            // Call IHP's delete binding, as we render the "delete" links after they have been executed.
        initDelete();

    }
};
$(document).on('ready turbolinks:load', initHandler);
```

With that script it's working for me now 👍

Reply:
I'm sorry! I forgot to update here that I've been working on a fix - https://github.com/amitaibu/ihp-sortable-tree/pull/2

Indeed, my problem was that I called things via Ajax, but didn't update it.

Thank you 🙏